source/PrintData.cpp

```diff
@@ -0,0 +1,473 @@
+/* PrintData.cpp
+Copyright (c) 2022 by warp-core
+
+Main function for Endless Sky, a space exploration and combat RPG.
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "PrintData.h"
+
+#include "GameData.h"
+#include "Outfit.h"
+#include "Ship.h"
+
+#include <iostream>
+#include <map>
+#include <set>
+
+using namespace std;
+
+
+
+void PrintData::Print(char *argv[])
+{
+	const char *const *it = argv + 2;
+	if(!*it)
+		return;
+	string arg = *it;
+	if(arg == "-s" || arg == "--ships")
+		Ships(argv);
+	else if(arg == "-w" || arg == "--weapons")
+		Weapons(argv);
+	else if(arg == "-e" || arg == "--engines")
+		PrintEngineStats();
+	else if(arg == "-p" || arg == "--power")
+		PrintPowerStats();
+	else if(arg == "-o" || arg == "--outfits")
+		Outfits(argv);
+	else if(arg == "-h" || arg == "--help")
+		Help();
+	cout.flush();
+}
+
+
+
+void PrintData::Help()
+{
+	cerr << "--printdata: follow with arguments to print diffferent data." << endl;
+	cerr << "    -s, --ships: prints table of various base stats of all ships (not variants)." << endl;
+	cerr << "        -d, --deterrences: prints ship names and deterrence values, not variants." << endl;
+	cerr << "        -v, --variants: includes variants." << endl;
+	cerr << "        -s, --sales: prints a list of ships and the outfitters they appear in." << endl;
+	cerr << "        -o, --old: the older version of the PrintShipTable() function." << endl;
```

Comment:
Do we need to keep the old version?

Reply:
It does things a little differently to my version. I don't see any harm in leaving it in and the information it provides might be useful to someone (someone thought it was useful enough to add in the first place).
I could update it to match the new format (commas instead of tabs), though.

Reply:
I agree with quyykk; there is no need to support printing the same data in two different formats; spreadsheet processor programs and linux-tools like sed perl and awk can be used to easily convert between the formats.

Removing duplication simplifies the source-code and makes it more maintainable. It also saves end-users to consider why there is an option called `"--old"`, because the term old is not really meaningful for someone that doesn't know the context of this parameter.

Reply:
I've updated the old one to use commas as well and changed the name to `PrintLoadedShipStats` since it uses `Ship::Attributes()` instead of `Ship::BaseAttributes()` like the other one (which has been renamed to `PrintBaseShipStats`) and also includes the heat and energy table you'd get in game, and the deterrence value of the ship as configured. It can optionally include variants.